models/iaf_wang_2002_exact.cpp

```diff
@@ -0,0 +1,570 @@
+ * Node initialization functions
+ * --------------------------------------------------------------------------- */
+
+void
+nest::iaf_wang_2002_exact::init_state_()
+{
+  assert( S_.state_vec_size == State_::s_NMDA_base );
+
+  double* old_state = S_.ode_state_;
+  S_.state_vec_size = State_::s_NMDA_base + 2 * ( S_.num_ports_ - SynapseTypes::GABA );
+  S_.ode_state_ = new double[ S_.state_vec_size ];
+
+  assert( S_.ode_state_ );
+
+  S_.ode_state_[ State_::V_m ] = old_state[ State_::V_m ];
+  S_.ode_state_[ State_::s_AMPA ] = old_state[ State_::s_AMPA ];
+  S_.ode_state_[ State_::s_GABA ] = old_state[ State_::s_GABA ];
+
+  for ( size_t i = State_::s_NMDA_base; i < S_.state_vec_size; ++i )
+  {
+    S_.ode_state_[ i ] = 0.0;
+  }
+
+  delete[] old_state;
+}
+
+void
+nest::iaf_wang_2002_exact::init_buffers_()
+{
+  B_.spikes_.resize( S_.num_ports_ );
+
+  for ( auto& sb : B_.spikes_ )
+  {
+    sb.clear(); // includes resize
+  }
+
+  B_.currents_.clear(); // includes resize
+
+  B_.weights_.resize( S_.num_ports_ - SynapseTypes::GABA + 1, 0.0 );
+
+  B_.logger_.reset(); // includes resize
+  ArchivingNode::clear_history();
+
+  if ( B_.s_ == nullptr )
+  {
+    B_.s_ = gsl_odeiv_step_alloc( gsl_odeiv_step_rkf45, S_.state_vec_size );
+  }
+  else
+  {
+    gsl_odeiv_step_reset( B_.s_ );
+  }
+
+  if ( B_.c_ == nullptr )
+  {
+    B_.c_ = gsl_odeiv_control_y_new( P_.gsl_error_tol, 0.0 );
+  }
+  else
+  {
+    gsl_odeiv_control_init( B_.c_, P_.gsl_error_tol, 0.0, 1.0, 0.0 );
+  }
+
+  if ( B_.e_ == nullptr )
+  {
+    B_.e_ = gsl_odeiv_evolve_alloc( S_.state_vec_size );
+  }
+  else
+  {
+    gsl_odeiv_evolve_reset( B_.e_ );
+  }
+
+  B_.sys_.function = iaf_wang_2002_exact_dynamics;
+  B_.sys_.jacobian = nullptr;
+  B_.sys_.dimension = S_.state_vec_size;
+  B_.sys_.params = reinterpret_cast< void* >( this );
+  B_.step_ = Time::get_resolution().get_ms();
+  B_.integration_step_ = Time::get_resolution().get_ms();
+
+  B_.I_stim_ = 0.0;
+}
+
+
+void
+nest::iaf_wang_2002_exact::pre_run_hook()
+{
+  // ensures initialization in case mm connected after Simulate
+  B_.logger_.init();
+
+  V_.RefractoryCounts = Time( Time::ms( P_.t_ref ) ).get_steps();
+  // since t_ref_ >= 0, this can only fail in error
+  assert( V_.RefractoryCounts >= 0 );
+}
+
+/* ---------------------------------------------------------------------------
+ * Update and spike handling functions
+ * --------------------------------------------------------------------------- */
+
+extern "C" inline int
+nest::iaf_wang_2002_exact_dynamics( double, const double ode_state[], double f[], void* pnode )
+{
+  // a shorthand
+  typedef nest::iaf_wang_2002_exact::State_ State_;
+
+  // get access to node so we can almost work as in a member function
+  assert( pnode );
+  nest::iaf_wang_2002_exact& node = *( reinterpret_cast< nest::iaf_wang_2002_exact* >( pnode ) );
+
+  // ode_state[] here is---and must be---the state vector supplied by the integrator,
```

Comment:
Why must it be? Explain.

Reply:
These comments are from the iaf_cond_exp model I copied, and are identical in all iaf_cond_* models. If we want to change it, would it be better to do it for all models in a separate PR?

Reply:
OK, then let's leave it.